File: backend/routers/interfaces/loopback.py
"""
Loopback Interface Configuration Endpoints

All loopback interface endpoints for VyOS configuration.
Loopback interfaces are limited to a single instance (lo) and support
address, description, ip source-validation, mirror, and redirect.
"""

import logging
from typing import Dict, List, Optional, Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field, ConfigDict
from starlette.concurrency import run_in_threadpool

from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from session_vyos_service import get_session_vyos_service
from routers.interfaces.interface_batch import (
    BatchRequest,
    VyOSResponse,
    run_interface_batch,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/loopback", tags=["loopback-interface"])


# ============================================================================
# Request / Response Models
# ============================================================================


class LoopbackInterfaceConfig(BaseModel):
    name: str
    type: str
    addresses: List[str] = Field(default_factory=list)
    description: Optional[str] = None
    ip_source_validation: Optional[str] = None
    mirror_ingress: Optional[str] = None
    mirror_egress: Optional[str] = None
    redirect: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class LoopbackInterfacesConfigResponse(BaseModel):
    interfaces: List[LoopbackInterfaceConfig] = Field(default_factory=list)
    total: int = 0


# ============================================================================
# Endpoints
# ============================================================================


@router.get("/capabilities")
async def get_capabilities(request: Request) -> Dict[str, Any]:
    """Return version-aware feature capabilities for loopback interfaces."""
    await require_read_permission(request, FeatureGroup.INTERFACES)
    service = get_session_vyos_service(request)
    from vyos_builders.interfaces.loopback import LoopbackInterfaceBuilderMixin
    builder = LoopbackInterfaceBuilderMixin(version=service.get_version())
    return builder.get_capabilities()


@router.get("/config", response_model=LoopbackInterfacesConfigResponse)
async def get_config(http_request: Request, refresh: bool = False) -> LoopbackInterfacesConfigResponse:
    """Get all loopback interface configurations from VyOS."""
    await require_read_permission(http_request, FeatureGroup.INTERFACES)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh)
        raw_config = full_config.get("interfaces", {}).get("loopback", {})

        from vyos_mappers.interfaces.loopback_versions import get_loopback_mapper
        mapper = get_loopback_mapper(service.get_version())
        parsed = mapper.parse_interfaces_of_type(raw_config)
        return LoopbackInterfacesConfigResponse(**parsed)
    except Exception:
        logger.exception("Unhandled error in get_config")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/batch", response_model=VyOSResponse)
async def batch_configure(http_request: Request, request: BatchRequest) -> VyOSResponse:
    """
    Configure a loopback interface using batch operations.

    **Supported operations:**
    | Operation | Value | Description |
    |-----------|-------|-------------|
    | `set_interface_description` | Yes | Set description |
    | `delete_interface_description` | No | Remove description |
    | `set_interface_address` | Yes | Add IP address (CIDR) |
    | `delete_interface_address` | Yes | Remove IP address |
    | `set_ip_source_validation` | Yes | Source validation (strict/loose/disable) |
    | `delete_ip_source_validation` | No | Remove source validation |
    | `set_mirror_ingress` | Yes | Mirror ingress to interface |
    | `delete_mirror_ingress` | No | Remove ingress mirror |
    | `set_mirror_egress` | Yes | Mirror egress to interface |
    | `delete_mirror_egress` | No | Remove egress mirror |
    | `set_redirect` | Yes | Redirect incoming packets |
    | `delete_redirect` | No | Remove redirect |
    | `delete_interface` | No | Delete entire interface |
    """
    await require_write_permission(http_request, FeatureGroup.INTERFACES)

    service = get_session_vyos_service(http_request)
    batch = service.create_loopback_batch()
    return run_interface_batch(service, batch, request)

File: frontend/src/lib/api/loopback.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface LoopbackInterface {
  name: string;
  type: string;
  addresses: string[];
  description: string | null;
  ip_source_validation: string | null;
  mirror_ingress: string | null;
  mirror_egress: string | null;
  redirect: string | null;
}

export interface LoopbackConfigResponse {
  interfaces: LoopbackInterface[];
  total: number;
}

export interface LoopbackCapabilities {
  version: string;
  features: Record<string, { supported: boolean; description: string }>;
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface LoopbackBatchOperation {
  op: string;
  value?: string;
}

// ============================================================================
// API Service
// ============================================================================

class LoopbackService {
  async getCapabilities(): Promise<LoopbackCapabilities> {
    return apiClient.get<LoopbackCapabilities>("/vyos/loopback/capabilities");
  }

  async getConfig(refresh: boolean = false): Promise<LoopbackConfigResponse> {
    return apiClient.get<LoopbackConfigResponse>("/vyos/loopback/config", {
      refresh: refresh.toString(),
    });
  }

  async refreshConfig(): Promise<VyOSResponse> {
    return apiClient.post("/vyos/config/refresh");
  }

  async batchConfigure(
    interfaceName: string,
    operations: LoopbackBatchOperation[]
  ): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/loopback/batch", {
      interface: interfaceName,
      operations,
    });
    await this.refreshConfig();
    return result;
  }

  async createInterface(config: {
    name: string;
    description?: string;
    addresses?: string[];
    ip_source_validation?: string;
    mirror_ingress?: string;
    mirror_egress?: string;
    redirect?: string;
  }): Promise<VyOSResponse> {
    const operations: LoopbackBatchOperation[] = [];

    if (config.description) operations.push({ op: "set_interface_description", value: config.description });
    if (config.addresses) {
      for (const addr of config.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }
    if (config.ip_source_validation) operations.push({ op: "set_ip_source_validation", value: config.ip_source_validation });
    if (config.mirror_ingress) operations.push({ op: "set_mirror_ingress", value: config.mirror_ingress });
    if (config.mirror_egress) operations.push({ op: "set_mirror_egress", value: config.mirror_egress });
    if (config.redirect) operations.push({ op: "set_redirect", value: config.redirect });

    return this.batchConfigure(config.name, operations);
  }

  async updateInterface(
    name: string,
    current: LoopbackInterface,
    updated: {
      description?: string | null;
      addresses?: string[];
      ip_source_validation?: string | null;
      mirror_ingress?: string | null;
      mirror_egress?: string | null;
      redirect?: string | null;
    }
  ): Promise<VyOSResponse> {
    const operations: LoopbackBatchOperation[] = [];

    // Simple string fields
    const stringFields: { key: keyof typeof updated; setOp: string; deleteOp: string; currentVal: string | null }[] = [
      { key: "description", setOp: "set_interface_description", deleteOp: "delete_interface_description", currentVal: current.description },
      { key: "ip_source_validation", setOp: "set_ip_source_validation", deleteOp: "delete_ip_source_validation", currentVal: current.ip_source_validation },
      { key: "mirror_ingress", setOp: "set_mirror_ingress", deleteOp: "delete_mirror_ingress", currentVal: current.mirror_ingress },
      { key: "mirror_egress", setOp: "set_mirror_egress", deleteOp: "delete_mirror_egress", currentVal: current.mirror_egress },
      { key: "redirect", setOp: "set_redirect", deleteOp: "delete_redirect", currentVal: current.redirect },
    ];

    for (const field of stringFields) {
      if (field.key in updated) {
        const newVal = updated[field.key] as string | null | undefined;
        if (newVal) {
          operations.push({ op: field.setOp, value: newVal });
        } else if (field.currentVal) {
          operations.push({ op: field.deleteOp });
        }
      }
    }

    // Array: addresses
    if (updated.addresses !== undefined) {
      for (const addr of current.addresses) {
        operations.push({ op: "delete_interface_address", value: addr });
      }
      for (const addr of updated.addresses) {
        operations.push({ op: "set_interface_address", value: addr });
      }
    }

    return this.batchConfigure(name, operations);
  }

  async deleteInterface(name: string): Promise<VyOSResponse> {
    return this.batchConfigure(name, [{ op: "delete_interface" }]);
  }
}

export const loopbackService = new LoopbackService();
